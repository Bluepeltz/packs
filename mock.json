{
  "id": "bluearchive",
  "title": "Blue Archive",
  "media": {
    "new": [
      {
        "format": "VIDEO_GAME",
        "popularity": 1000000,
        "externalLinks": [],
        "description": "Blue Archive is a role-playing game developed by Nexon Games, a subsidiary of Nexon. It was released in 2021 for Android and iOS, first in Japan by Yostar and worldwide later that year by Nexon. The game is free-to-play with gacha game mechanics as a means to obtain new characters.\n\nThe player and protagonist is a teacher who was summoned to the academy city of Kivotos by the president of the General Student Council, an extrajudicial committee governing the schools. Following her sudden disappearance, criminal activity rises around Kivotos, and the player is tasked by the remaining members of the council to resolve the issues and help search for the president.\n\n(Source: Wikipedia)",
        "characters": [
          {
            "role": "MAIN",
            "characterId": "hy_m"
          },
          {
            "role": "MAIN",
            "characterId": "idla"
          },
          {
            "role": "MAIN",
            "characterId": "yz-m"
          },
          {
            "role": "SUPPORTING",
            "characterId": "sqhb"
          },
          {
            "role": "SUPPORTING",
            "characterId": "xv9p"
          },
          {
            "role": "SUPPORTING",
            "characterId": "wknt"
          },
          {
            "role": "SUPPORTING",
            "characterId": "w2jf"
          },
          {
            "role": "SUPPORTING",
            "characterId": "i2xa"
          },
          {
            "role": "SUPPORTING",
            "characterId": "c8sg"
          },
          {
            "role": "SUPPORTING",
            "characterId": "oqu7"
          },
          {
            "role": "SUPPORTING",
            "characterId": "lv7c"
          },
          {
            "role": "SUPPORTING",
            "characterId": "ahzd"
          },
          {
            "role": "SUPPORTING",
            "characterId": "gy40"
          },
          {
            "role": "SUPPORTING",
            "characterId": "8ppg"
          },
          {
            "role": "SUPPORTING",
            "characterId": "enhq"
          },
          {
            "role": "SUPPORTING",
            "characterId": "blkq"
          },
          {
            "role": "SUPPORTING",
            "characterId": "i8ku"
          },
          {
            "role": "SUPPORTING",
            "characterId": "2hdl"
          },
          {
            "role": "SUPPORTING",
            "characterId": "inhn"
          },
          {
            "role": "SUPPORTING",
            "characterId": "osup"
          },
          {
            "role": "SUPPORTING",
            "characterId": "fkle"
          },
          {
            "role": "SUPPORTING",
            "characterId": "tzf4"
          },
          {
            "role": "SUPPORTING",
            "characterId": "dmff"
          },
          {
            "role": "SUPPORTING",
            "characterId": "yhxu"
          },
          {
            "role": "SUPPORTING",
            "characterId": "r2b5"
          },
          {
            "role": "SUPPORTING",
            "characterId": "g4mf"
          },
          {
            "role": "SUPPORTING",
            "characterId": "3eyw"
          }
        ],
        "id": "xqzn",
        "relations": [],
        "title": {
          "english": "Blue Archive",
          "alternative": ["BA"]
        },
        "type": "OTHER",
        "images": [
          {
            "url": "https://f005.backblazeb2.com/file/fable-packs/22TUW0Cu"
          }
        ]
      }
    ],
    "conflicts": []
  },
  "characters": {
    "new": [
      {
        "externalLinks": [],
        "name": {
          "english": "Takanashi Hoshino",
          "alternative": ["hoshino"]
        },
        "description": "Takanashi Hoshino is the former vice president and the only third-year of Abydos High School, and the current chairwoman for the Countermeasures Committee.\n\nSpeaking like a senile aunt, she enjoys playing around and napping more than work. Because of this, she usually gets scolded by the rest of the Committee. But, once the mission starts, she protects her comrades by fighting valiantly on the front lines.",
        "age": "17",
        "id": "hy_m",
        "media": [
          {
            "role": "MAIN",
            "mediaId": "xqzn"
          }
        ],
        "images": [
          {
            "url": "https://f005.backblazeb2.com/file/fable-packs/rtfbupny"
          }
        ],
        "gender": "Female"
      }
    ],
    "conflicts": []
  },
  "maintainers": ["228674702414053386", "839261971185664011"],
  "image": "https://f005.backblazeb2.com/file/fable-packs/dkihlBqc",
  "author": "Fable Community",
  "description": "A pack containing the various characters from the game Blue Archive",
  "conflicts": [
    "anilist:161049",
    "anilist:138416",
    "anilist:160589",
    "anilist:161050",
    "anilist:143177"
  ]
}
