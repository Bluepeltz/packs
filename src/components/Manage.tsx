import { Head } from '$fresh/runtime.ts';

import { createStyle } from 'flcss';

import Dialog from './Dialog.tsx';

import colors from '../utils/theme.ts';

import IconClose from 'icons/x.tsx';
import IconImage from 'icons/photo-plus.tsx';

import type { Schema } from './Dashboard.tsx';

import strings from '../../i18n/en-US.json' assert { type: 'json' };

const imagesTypes = [
  'image/png',
  'image/jpeg',
  'image/webp',
];

export default (props: { pack?: Schema.Pack }) => {
  const pack: Partial<Schema.Pack['manifest']> = props.pack?.manifest ?? {};

  const styles = createStyle({
    wrapper: {
      top: '0',
      left: '0',
      width: '100%',
      height: '100%',
      backgroundColor: colors.embedHighlight,
    },
    container: {
      display: 'grid',
      alignItems: 'center',
      gridTemplateColumns: '48px 1fr 32px',
      gridTemplateRows: '48px auto',
      gridTemplateAreas: '". . ." "_ _ _"',
      padding: '1.5em',
      gap: '2em',
    },
    image: {
      display: 'flex',
      position: 'relative',
      alignItems: 'center',
      justifyContent: 'center',
      boxSizing: 'border-box',

      overflow: 'hidden',
      borderRadius: '100%',
      height: '100%',
      width: '100%',

      border: `2px solid ${colors.grey}`,

      '> input': {
        width: '0',
      },

      '> img': {
        position: 'absolute',
        objectFit: 'cover',
        objectPosition: 'center',
        textIndent: '-100vh',
        height: '100%',
        width: '100%',
      },

      '> label': {
        position: 'absolute',
        width: '100%',
        height: '100%',
        cursor: 'pointer',
      },

      '> svg': {
        position: 'absolute',
        width: '28px',
        height: 'auto',
        color: colors.grey,
      },
    },
    close: {
      width: '100%',
      height: 'auto',
      cursor: 'pointer',
    },
    button: {
      position: 'fixed',
      backgroundColor: colors.discord,
      bottom: '0',
      right: '0',
      margin: '1.5em 3em',
    },
    body: {
      gridArea: '_',
      background: colors.red,
      height: '100px',
    },
  });

  return (
    <>
      <Head>
        <style>{styles.bundle}</style>
      </Head>

      <Dialog
        action={'back'}
        name={'manage'}
        visible={true}
        class={styles.names.wrapper}
      >
        <form
          method={'post'}
          action={'/api/publish'}
          encType={'multipart/form-data'}
          class={styles.names.container}
        >
          {/* used as a hack to send the existing pack json to the server with the post request */}
          <input type={'hidden'} name={'pack'} value={JSON.stringify(pack)} />

          <div class={styles.names.image}>
            <IconImage />
            <img src={pack.image ?? ''} data-image-cb={'pack_image'} />
            <label for={'pack_image'} />
            <input
              type={'file'}
              id={'pack_image'}
              name={'pack_image'}
              data-image={'pack_image'}
              accept={[...imagesTypes, 'image/gif'].join(',')}
            />
          </div>

          <input
            type={'text'}
            name={'pack_title'}
            placeholder={strings.packTitle}
            value={pack.title ?? pack.id}
          />

          <IconClose data-dialog-cancel={'manage'} class={styles.names.close} />

          {
            // <button class={styles.names.button} type='submit'>
            //   {strings.publish}
            // </button>
          }

          <div class={styles.names.body}>
            {/*  */}
          </div>
        </form>
      </Dialog>
    </>
  );
};
