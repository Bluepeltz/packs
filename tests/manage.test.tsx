import { renderToString } from 'preact-render-to-string';

import { assertSnapshot } from '$std/testing/snapshot.ts';

import { stub } from '$std/testing/mock.ts';

import Manage from '../src/components/Manage.tsx';

import mock from './mock.json' assert { type: 'json' };

Deno.test('<Manage pack={undefined}/>', async (test) => {
  const mathStub = stub(Math, 'random', () => 0);

  try {
    await assertSnapshot(
      test,
      renderToString(<Manage pack={undefined} />),
    );
  } finally {
    mathStub.restore();
  }
});

Deno.test('<Manage pack={pack}/>', async (test) => {
  const mathStub = stub(Math, 'random', () => 0);

  try {
    await assertSnapshot(
      test,
      // deno-lint-ignore no-explicit-any
      renderToString(<Manage pack={{ manifest: mock } as any} />),
    );
  } finally {
    mathStub.restore();
  }
});
