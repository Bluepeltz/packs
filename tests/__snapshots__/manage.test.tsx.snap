export const snapshot = {};

snapshot[`<Manage pack={undefined}/> 1`] = `'<div data-dialog-cb="manage" data-dialog-cb-action="back" data-dialog-cancel="manage" class="flcss-wrapper-"></div><div data-dialog-cb="manage" data-dialog-cb-action="back" class="flcss-container- flcss-wrapper-"><form method="post" action="/api/publish" encType="multipart/form-data" class="flcss-container-"><input type="hidden" name="pack" value="{}"/><div class="flcss-image-"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-photo-plus" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M15 8h.01"></path><path d="M12 20h-5a3 3 0 0 1 -3 -3v-10a3 3 0 0 1 3 -3h10a3 3 0 0 1 3 3v5"></path><path d="M4 15l4 -4c.928 -.893 2.072 -.893 3 0l4 4"></path><path d="M14 14l1 -1c.617 -.593 1.328 -.793 2.009 -.598"></path><path d="M16 19h6"></path><path d="M19 16v6"></path></svg><img src data-image-cb="pack_image"/><label for="pack_image"></label><input type="file" id="pack_image" name="pack_image" data-image="pack_image" accept="image/png,image/jpeg,image/webp,image/gif"/></div><input type="text" name="pack_title" placeholder="Pack Title"/><svg xmlns="http://www.w3.org/2000/svg" class="flcss-close-" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round" data-dialog-cancel="manage"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M18 6l-12 12"></path><path d="M6 6l12 12"></path></svg><div class="flcss-body-"></div></form></div>'`;

snapshot[`<Manage pack={pack}/> 1`] = `
\`<div data-dialog-cb="manage" data-dialog-cb-action="back" data-dialog-cancel="manage" class="flcss-wrapper-"></div><div data-dialog-cb="manage" data-dialog-cb-action="back" class="flcss-container- flcss-wrapper-"><form method="post" action="/api/publish" encType="multipart/form-data" class="flcss-container-"><input type="hidden" name="pack" value="{&quot;\$schema&quot;:&quot;https://raw.githubusercontent.com/ker0olos/fable/main/json/schema.json&quot;,&quot;id&quot;:&quot;utawarerumono&quot;,&quot;title&quot;:&quot;Utawarerumono Pack&quot;,&quot;description&quot;:&quot;Contains characters from Utawarerumono: Mask of Deception&quot;,&quot;author&quot;:&quot;Fable Community&quot;,&quot;image&quot;:&quot;https://media.tenor.com/n9YTIWj9H-gAAAAC/utawarerumono-kuon.gif&quot;,&quot;url&quot;:&quot;https://github.com/fable-community/fable-pack-utawarerumono&quot;,&quot;media&quot;:{&quot;conflicts&quot;:[&quot;anilist:21218&quot;],&quot;new&quot;:[{&quot;id&quot;:&quot;mask-of-deception&quot;,&quot;popularity&quot;:113000,&quot;title&quot;:{&quot;english&quot;:&quot;Utawarerumono: Mask of Deception&quot;,&quot;romaji&quot;:&quot;Utawarerumono: Itsuwari no Kamen&quot;,&quot;native&quot;:&quot;うたわれるもの 偽りの仮面&quot;,&quot;alternative&quot;:[&quot;Utawarerumono: The False Faces&quot;]},&quot;description&quot;:&quot;The only thing I can remember is coming to in the middle of an unfamiliar snowscape. Where that was or how I got there, I don't know. I couldn't even remember my own name. The memories just wouldn't come.\\
\\
I only had a moment, left at a loss, to ponder what to do — but fate seemed intent on furthering my misery as a giant, insect-like monster burst from the snow and ice to attack me.\\
\\
Naturally, I ran like hell. The creature cornered me anyway.\\
\\
That's when SHE appeared, as if from nowhere — the beautiful, mysterious girl who saved my life.\\
\\
She told me her name was Kuon. A beautiful, mysterious girl...\\
\\
...with a tail and ears like a beast's\\
\\
(Source: Atlus)&quot;,&quot;type&quot;:&quot;ANIME&quot;,&quot;images&quot;:[{&quot;url&quot;:&quot;https://s2.vndb.org/cv/68/45668.jpg&quot;}]}]},&quot;characters&quot;:{&quot;conflicts&quot;:[&quot;anilist:126823&quot;],&quot;new&quot;:[{&quot;id&quot;:&quot;kuon&quot;,&quot;gender&quot;:&quot;Female&quot;,&quot;name&quot;:{&quot;english&quot;:&quot;Kuon&quot;,&quot;romaji&quot;:&quot;Kuon&quot;,&quot;native&quot;:&quot;クオン&quot;,&quot;alternative&quot;:[&quot;kuon&quot;,&quot;Ku (Uta)&quot;]},&quot;description&quot;:&quot;An apothecary that left her hometown with ears and tail like an animal's. She's lively and intelligent, while somehow aloof.\\
\\
Separated from her homeland, she encounters Haku while traveling along leisurely.\\
\\
She gets excited when it comes to bathing and often ends her sentences with \\\\&quot;-かな～\\\\&quot; or, in English, \\\\&quot;I think\\\\&quot;.&quot;,&quot;media&quot;:[{&quot;mediaId&quot;:&quot;mask-of-deception&quot;,&quot;role&quot;:&quot;MAIN&quot;}],&quot;images&quot;:[{&quot;url&quot;:&quot;https://i.imgur.com/ClaapHd.png&quot;}]}]}}"/><div class="flcss-image-"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-photo-plus" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M15 8h.01"></path><path d="M12 20h-5a3 3 0 0 1 -3 -3v-10a3 3 0 0 1 3 -3h10a3 3 0 0 1 3 3v5"></path><path d="M4 15l4 -4c.928 -.893 2.072 -.893 3 0l4 4"></path><path d="M14 14l1 -1c.617 -.593 1.328 -.793 2.009 -.598"></path><path d="M16 19h6"></path><path d="M19 16v6"></path></svg><img src="https://media.tenor.com/n9YTIWj9H-gAAAAC/utawarerumono-kuon.gif" data-image-cb="pack_image"/><label for="pack_image"></label><input type="file" id="pack_image" name="pack_image" data-image="pack_image" accept="image/png,image/jpeg,image/webp,image/gif"/></div><input type="text" name="pack_title" placeholder="Pack Title" value="Utawarerumono Pack"/><svg xmlns="http://www.w3.org/2000/svg" class="flcss-close-" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round" data-dialog-cancel="manage"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M18 6l-12 12"></path><path d="M6 6l12 12"></path></svg><div class="flcss-body-"></div></form></div>\`
`;
